### PR TITLE
Issue/2778 Make Magda helm charts v3 charts & migration docs

### LINKED ISSUE #2778: Make our helm charts helm 3 chart
Our helm charts can be used by both helm 2 and helm 3.
But not making it 100% helm 3 and still support helm 2 make it hard to access all helm 3 features.
And it would require more effort to keep it compatible with helm 2 and 3.
Moreover, helm 2 support will end on Nov 13, 2020 (no security fixes after that).

## Acceptance Criteria
- A migration guide for existing helm2-based installations has been written and tested
    - This should allow a user to migrate with no data loss and minimal downtime (preferably < 5 minutes?)
    - This should take into account both pure-helm and terraform-based installations
    - This should take into account the process of getting the new helm command line tool, running helm 2to3 etc - we don't need to be copy-pasting existing documentation about migrating helm 2 installations but we do need to make sure we have a step-by-step document that links out to those guides where necessary and has been tested as working.
- The need for Helm 3 is now present in all the docs that tell you how to install helm (magda-config and building-and-running.md)
- Chart publishing has been moved over to Helm 3 (this will require some investigation because the way that helm3 does this isn't clear to us at the moment)

## Technical Notes
We can do this by:
- Change `apiVersion` in `Chart.yaml` to `v2`
  - May also specify `kubeVersion` so helm won't let users try Magda on unsupported k8s
- move `requirements.yaml`'s content to `Chart.yaml`
- Remove `requirements.lock` and commit `Chart.lock` instead
- Update document and specify the helm3 requirement for installing Magda
- We probably need a new charts repo (e.g. helm3.charts.magda.io or something like that)

### PR DESCRIPTION
**Pull request by t83714**
### What this PR does

Fixes #2778 

> Merged documentation PR: https://github.com/magda-io/magda/pull/2801 Please merge it first

Summary:
- Make Magda Helm chart requires Helm v3 & Kube API >= 1.14
- Fixed helm repo index generation using Helm 3 binary
- Added [v0.0.56-RC6-to-v0.0.57-0.md](https://github.com/magda-io/magda/blob/8979f9e2322c2d94b0c0dcc893f190680667fa35/docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md) doc
- Also, make sure `magda-config` users can deploy with terraform  and create PR here:
https://github.com/magda-io/magda-config/pull/10

### Test Repo index generation script

Before starting the test, install helm 2 (v2.16.3) using home brew and make the binary is accessible as `helm2` (helm v3 is accessible through helm)

Test Repo index generation script with local repo:
- Check out this PR branch `issue/2778`
- Delete the following directories files (if exists):
   - `deploy/helm/magda/charts`
   - `deploy/helm/magda/requirements.lock`
   - `deploy/helm/magda/requirements.yaml`
   - `deploy/helm/magda-core/charts`
   - `deploy/helm/magda-core/requirements.lock`
   - `deploy/helm/magda-core/requirements.yaml`
   - `deploy/helm/internal-charts/storage-api/charts`
   - `deploy/helm/internal-charts/storage-api/requirements.lock`
   - `deploy/helm/internal-charts/storage-api/requirements.yaml`
- Set AWS s3 access key & secret in env
- Run the followings:
```bash
cd deploy/helm
# we are going to serve local repo from this dir
mkdir chart-repo 
# download magda-charts bucket content
aws s3 sync s3://magda-charts/ chart-repo
mkdir -p sync_dir
mkdir -p index_dir
# download existing index file for index merging
if ! aws s3 cp s3://magda-charts/index.yaml index_dir/index.yaml; then echo "failed to copy current repo index" && exit 1; fi
helm package -d sync_dir openfaas
helm package -d sync_dir magda-core
helm package -d sync_dir magda
# generate index with merging
helm repo index --merge "index_dir/index.yaml" sync_dir
cp -f sync_dir/* chart-repo
```

Start local chart repo:
```bash
yarn global add local-web-server
cd chart-repo
ws
```

The local chart repo will be serving at http://127.0.0.1:8000

You can deploy Magda v0.0.56-RC6 with [v0.0.56-RC6 version `minikube-dev.yml`](https://github.com/magda-io/magda/blob/v0.0.56-RC6/deploy/helm/minikube-dev.yml) (with docker image change to `docker hub`)

```bash
helm2 --namespace magda upgrade --install --timeout 9999 --wait -f minikube-dev.yml magda magda --repo http://127.0.0.1:8000 --version 0.0.56-RC6
```

###  Test Migration Locally with Minikube

Download `helm_deployment_migration.sh` script from: https://gist.github.com/t83714/9317f8933223d3c4bd626128ee306208

Changes line 193 
from:
`helm get manifest "${helm_deployment_name}"`
to:
`helm2 get manifest "${helm_deployment_name}"`
as helm2 should be accessible via `helm2` for our test.

Run the following to build & push 0.0.57-0 docker images to local minikube (if you haven't done so):
```
yarn lerna run build --stream --concurrency=1 --include-dependencies
eval $(minikube docker)
yarn lerna run docker-build-local --stream --concurrency=1 --include-filtered-dependencies
```

Follow the instruction document [v0.0.56-RC6-to-v0.0.57-0.md](https://github.com/magda-io/magda/blob/768718d3a87bb33f9d0a4dc24fc08b51e3a65d79/docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md) for migration except `Step 6`.

You can use the following to install the chart from your local repo instead:
```bash
helm --namespace magda upgrade --install --timeout 9999s --wait -f deploy/helm/minikube-dev.yml magda magda --repo http://127.0.0.1:8000 --version 0.0.57-0 --set global.connectors.includeInitialJobs=false
```

### Checklist

-   [x] unit tests aren't applicable
-   [x] I've updated CHANGES.md with what I changed.
-   [x] I've linked this PR to an issue in ZenHub (core dev team only)

### CODE REVIEW

File: docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+helm repo add stable https://kubernetes-charts.storage.googleapis.com
+```
+
+to add the Magda helm chart repo.
+
+Run `helm upgrade` to upgrade your release:
+
+```bash
+helm --namespace [your magda deploy namespace] upgrade --install --timeout 9999s --wait -f [path to your config value file] magda magda --repo https://charts.magda.io --version 0.0.57-0
+```
```
Comment (AlexGilleran) [inline docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md:159]: As discussed, I'm getting 

```
Error: UPGRADE FAILED: rendered manifests contain a new resource that already exists. Unable to continue with update: existing resource conflict: namespace: default, name: es-ds, existing_kind: apps/v1, Kind=DaemonSet, new_kind: apps/v1, Kind=DaemonSet
```
:(

File: docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+Magda v0.0.57-0 introduced two breaking changes to Magda helm charts:
+
+-   Deprecated k8s APIs used in Helm charts are upgraded to support k8s Api 1.16. You can find the list of deprecated APIs removed since k8s 1.16 from [here](https://kubernetes.io/blog/2019/07/18/api-deprecations-in-1-16/).
+    -   The problem with it is that Helm recognises release objects by k8s APIs. Therefore, for example, a previous deployed `DaemonSet` under k8s API `extensions/v1beta1` in version v0.0.56-RC6 will be considered as a different object as it's now deployed through k8s API `apps/v1` in version v0.0.57-0. During the upgrade, helm will attempt to create another `DaemonSet` under k8s API `apps/v1` but ends up an `DaemonSet exists` error.
+-   Since version v0.0.57-0, you can only deploy Magda with Helm 3. If you try to deploy Magda with Helm 2, you will get a `Error: apiVersion 'v2' is not valid. The value must be "v1"` error.
+
+#### Migration Path Overview
```
[the remark this answers is not in the document]
Comment (t83714) [reply docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md:15]: Removed the `Migration Path Overview` section 👍

File: docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+#### Readme before migration
+
+> WARNING: All data migrations carry a level of risk. This migration is no different. You should be aware of any risks specific to your environment and prepare a data migration strategy for your needs.
+
+Here are some suggestions to mitigate against potential risks during migration:
+
+-   Perform a data backup of the following:
+    -   Helm v2 home folder.
+    -   Release data from the cluster. Refer to [How Helm Uses ConfigMaps to Store Data](http://technosophos.com/2017/03/23/how-helm-uses-configmaps-to-store-data.html) for details on how Helm v2 store release data in the cluster.
```
[the remark this answers is not in the document]
Comment (t83714) [reply docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md:36]: It's copied from `2to3` plugin readme~ I think I probably just remove it 😄

File: docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+-   Prerequisites required for `helm_deployment_migration.sh`:
+    -   Create `repos` directory in `Your Home Directory`: `cd ~ && mkdir repos`.
+        -   We will save required `helm` & `protobuf` repos to this folder
+    -   Helm 2 protobuf schema: you need to clone [helm github repo](https://github.com/helm/helm.git) to `[Your Home Directory]/repos/helm` with the following commands:
+        -   `cd ~/repos`
+        -   `git clone https://github.com/helm/helm.git`
+    -   Protbuf base schema: you need to clone [google's protbuf github repo](https://github.com/google/protobuf) to `[Your Home Directory]/repos/protobuf` with the following commands:
+        -   `cd ~/repos`
+        -   `git clone https://github.com/google/protobuf`
+    -   `kubectl`: You can install with [homebrew](https://brew.sh/) (if you haven't install one)
```
[the remark this answers is not in the document]
Comment (t83714) [reply docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md:53]: It's for covering the terraform users -- they don't have to have the `kubectl` installed to install the Magda

File: docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        -   `brew install kubectl`
+    -   `protoc` binary: You can install with [homebrew](https://brew.sh/) (if you haven't install one)
```
[the remark this answers is not in the document]
Comment (t83714) [reply docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md:55]: Good point 👍  Added the link in

File: docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md (hunk 1 above)
Comment (t83714) [reply docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md:159]: I missed one step:
```bash
kubectl apply -f https://raw.githubusercontent.com/magda-io/magda/master/deploy/helm/openfaas/crds/crd.yaml
```
in the doc --- have added that step in.
If your cluster already in error status, you can try helm v3.2.0 as it handles errors status better.

File: docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        -   `brew install protobuf`
+
+> NOTE: You need to checkout the corresponding helm v2 branch of your tiller deployment in the helm repo that you've just cloned, e.g. v2.16.1 Otherwise the headers won't be there (they were with helm V3)
+
+To find out your helm tiller version, you can:
+
+```bash
+helm verison
```
[the remark this answers is not in the document]
Comment (t83714) [reply docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md:41]: 👍 fixed

File: docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (AlexGilleran) [reply docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md:30]: ah right :+1:

File: docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md [hunk 7]
```diff
@@ -0,0 +1,145 @@
+### Migrate from Magda v0.0.56-RC6 to v0.0.57-0
```
Comment (AlexGilleran) [inline docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md:1]: I just realised that I had one more comment that I didn't put in before - can we call this just `0.0.56 to 0.0.57`, without the extra versions? :)

Comment (t83714) [reply docs/docs/migration/v0.0.56-RC6-to-v0.0.57-0.md]: The file name is changed & all relevant keywords have also been changed in the document

**Comment by t83714** [comment]
@AlexGilleran I think it's ready for another look now~

**Pull request merged**